### DISCUSSION TITLE
Using TabularPredictor with more than one eval_metric?

### DISCUSSION THREAD

**Discussion by TonciG** [autogluon/autogluon#5655] [Q&A] [answered]
Hi,

Does anyone have an idea on how to (and whether it is possible) use/extend TabularPredictor in such situation?

For example, my target variable is customer churn but I would like to optimise a classifier that has both high recall for positive class (1 or churn = yes) as well as high recall for negative class (0 or churn = no). Is the [multi-label prediction example](https://auto.gluon.ai/stable/tutorials/tabular/index.html) a possible way forward?

Thanks!

**Comment by cookesan** [comment, marked as answer]
`TabularPredictor` optimizes one `eval_metric` for model selection. The constructor docs describe `eval_metric` as the metric AutoGluon uses to tune hyperparameters, early stopping, ensemble weights, and the default "best" model.

You can still report additional metrics after training:

```python
predictor = TabularPredictor(
    label="churn",
    eval_metric="roc_auc",
).fit(train_data)

predictor.evaluate(test_data, auxiliary_metrics=True)

predictor.leaderboard(
    test_data,
    extra_metrics=["accuracy", "f1", "precision", "recall"],
)
```

If the business objective really is "optimize a combination of metrics", make that combination a single custom scorer and pass it as `eval_metric`. Otherwise AutoGluon needs one metric to decide which validation score is better.

**Comment by TonciG** [comment]
Thank your response and a great library you've helped to create.
I've managed to optimise a TabularPredictor for both positive class (1 or churn = yes) as well as negative class (0 or churn = no). Used F1 instaed of Recall. 
Just used your class (only changing consider_labels_correlation = False) of [Multil-label Predictor](https://auto.gluon.ai/stable/tutorials/tabular/advanced/tabular-multilabel.html) and created two custom metrics (f1_yes and f1_no) with [autogluon's make_corer](https://auto.gluon.ai/stable/tutorials/tabular/advanced/tabular-custom-metric.html) and [scikit-learn's f1 scorer](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.f1_score.html), see below:
from functools import partial
from sklearn.metrics import f1_score
f1_yes_scorer = make_scorer(name = 'f1_yes',
                                 score_func = partial(f1_score, pos_label = 1),
                                 optimum = 1,
                                 greater_is_better = True,
                                 needs_class = True,
                                 needs_proba = False)

f1_no_scorer = make_scorer(name = 'f1_no',
                                score_func = partial(f1_score, pos_label = 1),
                                optimum = 1,
                                greater_is_better = True,
                                needs_class = True,
                                needs_proba = False)

The final thing I did was to change my dataset so it can be used to train a Multimetric Predictor by creating two target columns (Churn_yes and Churn_No). Need to make sure the values in these columns are opposite so they can be used to maximise f1_yes and f1_no metrics.
Everything works fine now and I am able to train and make predictions with a single Multimetric Tabular Predictor.
Hope this make sense to you?
